addons/udes_stock/models/stock_move_line.py

```diff
@@ -366,10 +360,152 @@ def prepare(
+                domain = [
+                    ("product_id", "=", product.id),
+                    ("location_id", "child_of", location.ids),
+                    ("lot_id", "in", [lot_id.id for lot_id in lot_ids_swapped_in]),
+                ]
+                quants = Quant.search(domain).with_context(prefetch_fields=False)
+                quants.read(["quantity", "reserved_quantity"], load="_classic_write")
+                if quantity_swapping > sum(quants.mapped("quantity")):
+                    raise UserError(
+                        "You have picked more stock than is available on the system. Speak to your manager."
+                    )
+
+                # Generator expressions to easily iterate over
+                mls_need_changing = (ml for ml in mls_need_changing)
+                lot_ids_swapped_in = (lot for lot in lot_ids_swapped_in)
+
+                amount_to_be_swapped = quantity_swapping
+                if is_serial:
+                    # Want to update the movelines on each serial tracked product
+                    # until all the quantity scanned in is zero
+                    qty_done = 1
+                    while quantity_swapping > 0:
+                        ml, lot_id = next(mls_need_changing), next(lot_ids_swapped_in)
+                        prod_dict = {
+                            "product_uom_qty": qty_done,
+                            "qty_done": qty_done,
+                            "lot_name": lot_id.name,
+                            "lot_id": lot_id.id,
+                        }
+                        prod_dict.update(vals)
+                        res[ml] = prod_dict
+                        quantity_swapping -= qty_done
+                else:
+                    qty_done = quantity_swapping
+                    # Update the first lot with all scanned in lot information
+                    ml, lot_id = next(mls_need_changing), next(lot_ids_swapped_in)
+                    reserved_quantity = ml.product_uom_qty
+                    prod_dict = {
+                        "product_uom_qty": qty_done,
+                        "qty_done": qty_done,
+                        "lot_name": lot_id.name,
+                        "lot_id": lot_id.id,
+                    }
+                    prod_dict.update(vals)
+                    res[ml] = prod_dict
+                    quantity_swapping -= reserved_quantity
+                    # If the lot quantity covers multiple movelines then delete these movelines
+                    # Unless it has partially picked some moveline then update the reserved qty on that moveline
+                    while quantity_swapping > 0:
+                        ml = next(mls_need_changing)
+                        reserved_quantity = ml.product_uom_qty
+                        quantity_swapping -= reserved_quantity
+                        if quantity_swapping >= 0:
+                            ml.unlink()
```

Comment:
I think we should be able to cope without unlinking by just preserving original `product_uom_qty` and updating the `lot_id` of the move lines.